cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,307 @@
+        material_id_metadata_dict = dict()
+        for node in nodes_to_check:
+            if node is not None:
+                for material_id, node in node.material_map.items():
+                    if material_id not in material_id_metadata_dict:
+                        material_id_metadata_dict[material_id] = node
+
+        return material_id_metadata_dict
+
+    #
+    # Gets MaterialNode for the given extruder and machine with the given material name.
+    # Returns None if:
+    #  1. the given machine doesn't have materials;
+    #  2. cannot find any material InstanceContainers with the given settings.
+    #
+    def getMaterialNode(self, machine_definition_id: str, variant_name: Optional[str], diameter: float, root_material_id: str) -> Optional["InstanceContainer"]:
+        # round the diameter to get the approximate diameter
+        rounded_diameter = str(round(diameter))
+        if rounded_diameter not in self._diameter_machine_variant_material_map:
+            Logger.log("i", "Cannot find materials with diameter [%s] (rounded to [%s]) for root material id [%s]",
+                       diameter, rounded_diameter, root_material_id)
+            return None
+
+        # If there are variant materials, get the variant material
+        machine_variant_material_map = self._diameter_machine_variant_material_map[rounded_diameter]
+        machine_node = machine_variant_material_map.get(machine_definition_id)
+        variant_node = None
+
+        # Fallback for "fdmprinter" if the machine-specific materials cannot be found
+        if machine_node is None:
+            machine_node = machine_variant_material_map.get(self._default_machine_definition_id)
+        if machine_node is not None and variant_name is not None:
+            variant_node = machine_node.getChildNode(variant_name)
+
+        # Fallback mechanism of finding materials:
+        #  1. variant-specific material
+        #  2. machine-specific material
+        #  3. generic material (for fdmprinter)
+        nodes_to_check = [variant_node, machine_node,
+                          machine_variant_material_map.get(self._default_machine_definition_id)]
+
+        material_node = None
+        for node in nodes_to_check:
+            if node is not None:
+                material_node = node.material_map.get(root_material_id)
+                if material_node:
+                    break
+
+        return material_node
+
+    #
+    # Used by QualityManager. Built-in quality profiles may be based on generic material IDs such as "generic_pla".
+    # For materials such as ultimaker_pla_orange, no quality profiles may be found, so we should fall back to use
+    # the generic material IDs to search for qualities.
+    #
+    # This function returns the generic root material ID for the given material type, where material types are "PLA",
+    # "ABS", etc.
+    #
+    def getFallbackMaterialId(self, material_type: str) -> str:
```

Comment:
[Will document this]

It's not exactly the same as what `getRootMaterialId()` would imply. In the current naming scheme, what `getRootMaterialId(str)` would imply is that if we give `generic_abs_ultimaker3_AA_0.4` as the argument, it will return `generic_abs`. What the "Fallback" here means can be explained in the example below:

Suppose a machine says my preferred material is "filo3d_pla" but its extruders only use 2.85mm materials, then there won't be any material available for this machine because of this incorrect setup in its definition file. A fallback would be to fetch a generic material of the same type "PLA" as "filo3d_pla", and in this case it will be "generic_pla". This function is intended to get a generic fallback material for the given material type.

I will add this doc to the code and rename the function to be more descriptive.

Reply:
Documented and renamed in 50c61c993.